tree/30_generic_methods/user_secondary_group.cf

```diff
@@ -0,0 +1,161 @@
+
+bundle agent user_secondary_groups(login, groups, force) {
+  vars:
+      "class_prefix"              string => canonify("user_secondary_groups_${login}");
+      "group_list"                     slist => string_split("${groups}", ",", "999999");
+      "subcall_prefix[${group_list}]" string => canonify("user_group_${login}_${group_list}");
+      "prefixes"                       slist => { "${class_prefix}",
+                                                  "${report_data.method_id}"
+                                                };
+      "c_id"                          string => canonify("${report_data.method_id}");
+
+
+  defaults:
+      "force" string => "false", if_match_regex => "";
+
+  classes:
+      "user_exists"         expression => userexists("${login}");
+      "group_not_found"            not => groupexists("${group_list}");
+      "purge_other_groups"  expression => strcmp("${force}", "true");
+      "should_check"               and => { "user_exists", "!group_not_found" };
+
+      "should_report"       expression => "${report_data.should_report}";
+
+    any::
+      "pass3" expression => "pass2";
+      "pass2" expression => "pass1";
+      "pass1" expression => "any";
+
+
+    pass3::
+      "error_subcall"    expression => "${subcall_prefix[${group_list}]}_error";
+      "repaired_subcall" expression => "${subcall_prefix[${group_list}]}_repaired";
+      "success_subcall"  expression => and( "${subcall_prefix[${group_list}]}_kept",
+                                            not("error_subcall|repaired_subcall")
+                                          );
+
+      # ${c_id}_{error|repaired|success} are already defined if in enforce
+      # There are redefined here for clarity
+      "${c_id}_error"    expression => or( "!user_exists",
+                                           "group_not_found",
+                                           "error_subcall",
+                                           "${c_id}_error"
+                                         );
+
+      "${c_id}_repaired" expression => and( "user_exists",
+                                            "!group_not_found",
+                                            or( "repaired_subcall",
+                                                "${c_id}_repaired"
+                                              )
+                                          );
+
+      "${c_id}_success"  expression => and( "user_exists",
+                                            "!group_not_found",
+                                            or( "success_subcall",
+                                                "${c_id}_success"
+                                              )
+                                          );
+
+  users:
+    pass2::
+      "${login}"
+        policy => "present",
+        groups_secondary => { @{group_list} },
+        classes => classes_generic_two("${class_prefix}", "${report_data.method_id}"),
+        if => and( "user_exists",
```

Comment:
this should be only if should_check and purge_orhe_groups, isn't it?

Reply:
Yes, but sould_check should not be there. It was use in the early draft of the method, will remove it.

Reply:
fixed